titus-server-master/src/main/java/com/netflix/titus/master/jobmanager/endpoint/v3/grpc/DefaultJobManagementServiceGrpc.java

```diff
@@ -345,38 +349,24 @@ public void findTask(TaskId request, StreamObserver<Task> responseObserver) {
-                    })
-                    .orElseThrow(() -> JobManagerException.jobNotFound(request.getJobId()));
-
+            com.netflix.titus.api.jobmanager.model.job.Job<ServiceJobExt> serviceJobExtJob = verifyAndGetServiceJob(request.getJobId());
             com.netflix.titus.api.jobmanager.model.job.Capacity newCapacity = V3GrpcModelConverters.toCoreCapacity(request.getCapacity());
+            sanitizeAndUpdateJobCapacity(request.getJobId(), newCapacity, serviceJobExtJob.getJobDescriptor(), callMetadata, responseObserver);
+        });
+    }
 
-            Set<ValidationError> violations = CollectionsExt.merge(
-                    entitySanitizer.validate(newCapacity),
-                    validateCustomJobLimits(JobFunctions.changeServiceJobCapacity(job.getJobDescriptor(), newCapacity))
-            );
-            if (!violations.isEmpty()) {
-                safeOnError(logger, TitusServiceException.invalidArgument(violations), responseObserver);
-                return;
-            }
+    @Override
+    public void updateJobCapacityWithOptionalAttributes(JobCapacityUpdateWithOptionalAttributes request, StreamObserver<Empty> responseObserver) {
+        execute(callMetadataResolver, responseObserver, callMetadata -> {
+            com.netflix.titus.api.jobmanager.model.job.Job<ServiceJobExt> serviceJobExtJob = verifyAndGetServiceJob(request.getJobId());
+            Capacity.Builder newCapacityBuilder = serviceJobExtJob.getJobDescriptor().getExtensions().getCapacity().toBuilder();
 
-            authorizeJobUpdate(callMetadata, request.getJobId())
-                    .concatWith(jobOperations.updateJobCapacityReactor(request.getJobId(), newCapacity, callMetadata))
-                    .subscribe(
-                            nothing -> {
-                            },
-                            e -> safeOnError(logger, e, responseObserver),
-                            () -> {
-                                responseObserver.onNext(Empty.getDefaultInstance());
-                                responseObserver.onCompleted();
-                            }
-                    );
+            final JobCapacityWithOptionalAttributes jobCapacityWithOptionalAttributes = request.getJobCapacityWithOptionalAttributes();
+            Evaluators.acceptIfTrue(jobCapacityWithOptionalAttributes.hasDesired(), valueAccepted -> newCapacityBuilder.withDesired(jobCapacityWithOptionalAttributes.getDesired().getValue()));
+            Evaluators.acceptIfTrue(jobCapacityWithOptionalAttributes.hasMax(), valueAccepted -> newCapacityBuilder.withMax(jobCapacityWithOptionalAttributes.getMax().getValue()));
+            Evaluators.acceptIfTrue(jobCapacityWithOptionalAttributes.hasMin(), valueAccepted -> newCapacityBuilder.withMin(jobCapacityWithOptionalAttributes.getMin().getValue()));
+
+            sanitizeAndUpdateJobCapacity(request.getJobId(), newCapacityBuilder.build(), serviceJobExtJob.getJobDescriptor(), callMetadata, responseObserver);
```

Comment:
I think this is still prone to race conditions, if multiple partial updates are being executed concurrently, they will undo each others updates. I think each partial update should be done inside `BasicServiceJobActions.updateJobCapacityAction(...)` (used by `DefaultV3JobOperations#updateJobCapacity()`) as an atomic operation. Then all the interfaces from `V3JobOperations` and below can only accept `capacityWithOptionalAttributes`, and the original call can be implemented on top of it (passing down all three min, max, desired).

Reply:
One thing to note is that validating that `min <= desired <= max` will need to be made inside the same atomic operation, in the reconciliation engine

Reply:
Yeah that's what I was concerned about since I wasn't really sure how reconciliation engine was serializing model actions. Good catch.